cove/lib/common.py

```diff
@@ -342,3 +371,112 @@ def get_json_data_deprecated_fields(schema_name, schema_url, json_data):
+    '''
+    if codelist_paths is None:
+        codelist_paths = {}
+
+    if schema_url:
+        obj = get_schema_data(schema_url, schema_name)
+
+    for prop, value in obj['properties'].items():
+        if current_path:
+            path = current_path + (prop,)
+        else:
+            path = (prop,)
+
+        if "codelist" in value and path not in codelist_paths:
+            codelist_paths[path] = (value['codelist'], value.get('openCodelist', False))
+
+        if value.get('type') == 'object':
+            _get_schema_codelist_paths(None, None, value, path, codelist_paths)
+        elif value.get('type') == 'array' and value['items'].get('properties'):
+            _get_schema_codelist_paths(None, None, value['items'], path, codelist_paths)
+
+    return codelist_paths
+
+
+@functools.lru_cache()
+def _load_codelists(codelist_url, unique_files):
+    codelists = {}
+    for codelist_file in unique_files:
+        try:
+            response = requests.get(codelist_url + codelist_file)
+            response.raise_for_status()
+            reader = csv.DictReader(line.decode("utf8") for line in response.iter_lines())
+            codelists[codelist_file] = {}
+            for record in reader:
+                code = record.get('Code') or record.get('code')
+                title = record.get('Title') or record.get('Title_en')
+                codelists[codelist_file][code] = title
+
+        except requests.exceptions.RequestException:
+            codelists = {}
+            break
+
+    return codelists
+
+
+def _generate_data_path(json_data, path=()):
+    for key, value in json_data.items():
+        if not value:
+            continue
+        if isinstance(value, list):
+            if isinstance(value[0], dict):
+                for num, item in enumerate(value):
+                    yield from _generate_data_path(item, path + (key, num))
+            else:
+                yield path + (key,), value
+        elif isinstance(value, dict):
+            yield from _generate_data_path(value, path + (key,))
+        else:
+            yield path + (key,), value
+
+
+def get_additional_codelist_values(schema_name, schema_url, codelist_url, json_data):
+    if not codelist_url:
+        return {}
+    codelist_schema_paths = _get_schema_codelist_paths(schema_name, schema_url)
+    unique_files = frozenset(value[0] for value in codelist_schema_paths.values())
+    codelists = _load_codelists(codelist_url, unique_files)
+
+    additional_codelist_values = {}
+    for path, values in _generate_data_path(json_data):
+        if not isinstance(values, list):
+            values = [values]
+
+        path_no_num = tuple(key for key in path if isinstance(key, str))
+
+        if path_no_num not in codelist_schema_paths:
+            continue
+
+        codelist, isopen = codelist_schema_paths[path_no_num]
+
+        codelist_values = codelists.get(codelist)
+        if not codelist_values:
+            continue
```

Comment:
This is because we don't support codelists for extensions yet. Skipping here means we at least don't fail.